### PR TITLE
Camera abstraction

### PR DESCRIPTION
**Pull request by diegoasua**
An abstraction of the digital camera to give room to others. This PR changes and breaks down code directly controlling the CMOS camera. Beware chances are high this will need a bit of trial and error in the setup (many things could have gone wrong during this process including invalidating auto-pickling from the Python interpreter, which is key for multiprocessing). Briefly,

- Hamamtsu ORCA-flash API was cleaned and broke down into the SDK component and the pure API.
- A BasicCamera object was created in `..\sashimi\hardware\cameras\__init__.py`
- `camera.py` was adapted to these changes

### CODE REVIEW

**Comment by diegoasua** [comment]
What is NOXSESSION="lint" and why does it fail this build? I see it is failing even in master?

**Comment by vilim** [comment]
Nice! Consider making `@property` decorators for the camera settings used by sashimi (so you would just do e.g. `camera.exposure = 0.010`). Another minor point, the hamatsu code here is more accurately called a wrapper rather than SDK or API

**Comment by vilim** [comment]
This PR should also remove the MockCameraProcess and have a MockCamera instead

**Comment by diegoasua** [comment]
> Nice! Consider making `@property` decorators for the camera settings used by sashimi (so you would just do e.g. `camera.exposure = 0.010`). Another minor point, the hamatsu code here is more accurately called a wrapper rather than SDK or API

Not sure how to make the decorators for the settings :) Should each setting have its own method inside a broader class `CameraSettings`? Should it have `property.setter` decorators? Should this go in another function? Sorry I just don´t know about this.

**Comment by vilim** [comment]
Yes. There should be an `AbstractCamera` that defines all the things that need to be implemented for a camera (have a look [here](https://github.com/portugueslab/sashimi/blob/scanningabstraction/sashimi/hardware/scanning/interface.py) ), and then the HamamatsuCamera redefines it.

**Comment by vilim** [comment]
Also, I don't think you need to overspecify the camera model in the class names (the DCAM API works for all Hamamatsu cameras)

**Comment by diegoasua** [comment]
Is this more or less what you were referring to? `AbstractCameraInterface` contains the properties and properties.setters. `BasicCamera` inherits it and adds the rest of the high-level conviniency functions for `camera.py`.

File: sashimi/camera.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from multiprocessing import Process, Queue
 from enum import Enum
 from arrayqueues.shared_arrays import ArrayQueue
-from sashimi.hardware.cameras.Hamamatsu_ORCA_flash_wrapper import HamamatsuOrcaFlashCamera
+from sashimi.hardware.cameras.Hamamatsu_wrapper import HamamatsuCamera
```
Comment (vilim) [inline sashimi/camera.py:4]: lower case please, even Hamamatsu does not have the honor of breaking Python noventions :)

File: sashimi/hardware/cameras/__init__.py [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
+from sashimi.hardware.cameras.interface import AbstractCameraInterface
 
 conf = read_config()
 
 
-class BasicCamera:
+class BasicCamera(AbstractCameraInterface):
     def __init__(self):
+        super().__init__()
         self.encoding = "utf-8"
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:12]: What is this?

**Review by vilim — CHANGES_REQUESTED** [review]
Good! Now just putting the Hamamatsu interface and this is good to go! (or a fake camera, as you wish, or both, but at least one needs to be working)

**Comment by vilim** [comment]
I must admit I don't get the point of having both a BasicCamera and an AbstractCameraInterface, I think having the abstract interface is enough (more complicated structure will be necessary if cameras with more features want to be added)

File: sashimi/hardware/cameras/__init__.py (hunk 2 above)
Comment (diegoasua) [reply sashimi/hardware/cameras/__init__.py:12]: I don´t really know. It was part of the old wrapper and not even sure what it does

File: sashimi/hardware/cameras/__init__.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        frames = []
+        is_there_frame = np.random.random_sample(1)
+        if is_there_frame < (1 / self.exposure_time):
+            frame = np.random.randint(
+                0, 65534, size=self.frame_shape, dtype=np.uint16
+            )
+            frames.append(frame)
+        return frames
+
+    def get_internal_parameters(self, parameters, *args, **kwargs):
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:176]: the `frame_shape` and a `frame_rate` property should handle this

File: sashimi/hardware/cameras/__init__.py (hunk 2 above)
Comment (vilim) [reply sashimi/hardware/cameras/__init__.py:11]: It definitely doesn't do anything for any camera, just remove it.

File: sashimi/hardware/cameras/__init__.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
+from sashimi.config import read_config
+import time
+import numpy as np
+
+conf = read_config()
+
+
+class BasicCamera:
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:8]: rename this to AbstractCamera because it doesn't do anything (classes that are called Abstract need the inherited ones to implement things, whereas the name Basic implies it already works in a simple case)

File: sashimi/hardware/cameras/__init__.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    def __init__(self):
+        super().__init__()
+        self.encoding = "utf-8"
+        self.camera_id = conf["camera"]["id"]
+        self.image_height = 2048
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:13]: it is inconsistent to have image split into height and weight and the frame shape as a tuple. I would instead suggest having a property `sensor_resolution` that is a tuple

File: sashimi/hardware/cameras/__init__.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        self.image_width = 2048
+        self.frame_shape: tuple = (1024, 1024)
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:15]: since frame_shape is affected by property getters and setters, the python convention is to call it with an underscore (_frame_shape), see e.g. the paragraph [here](https://pep8.org/#designing-for-inheritance)

File: sashimi/hardware/cameras/__init__.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        pass
+
+
+class MockCamera(BasicCamera):
+    def __init__(self):
+        super().__init__()
+        self.internal_frame_shape: tuple = (1024, 1024)
+        self.internal_exposure_time = 60
+
+    def apply_parameters(self, parameters, *args, **kwargs):
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:140]: this is redundant with the setters

File: sashimi/hardware/cameras/__init__.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        return BasicCamera()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+
+class MockCamera(BasicCamera):
+    def __init__(self):
+        super().__init__()
+        self.internal_frame_shape: tuple = (1024, 1024)
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:137]: instead of `internal*`, use the python convention of prefixing with underscore

**Review by vilim — CHANGES_REQUESTED** [review]
almost there with the MockCamera!

File: sashimi/hardware/cameras/__init__.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+        self._frame_rate = 1 / self._exposure_time
+
+    # TODO: Modify camera.py to access properties directly and not on-top functions like apply_parameters()
+
+    @property
+    def exposure_time(self):
+        return self._exposure_time
+
+    @exposure_time.setter
+    def exposure_time(self, parameters):
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:171]: it is slightly wrong: you will not pass a parameters object, but directly exposure time as a number

File: sashimi/hardware/cameras/__init__.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+    @property
+    def frame_shape(self):
+        return self._frame_shape
+
+    @frame_shape.setter
+    def frame_shape(self, parameters):
+        subarray = parameters.subarray
+        # quantizing the ROI dims in multiples of 4
+        subarray = [min((i * parameters.binning // 4) * 4, 2048) for i in subarray]
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:182]: it's unnecessary to do this for the mock camera

File: sashimi/camera.py [hunk 11]
```diff
@@ -153,6 +150,13 @@ def camera_loop(self):
             except Empty:
                 pass
 
+    def update_parameters(self):
+        self.parameters = self.new_parameters
+        self.camera.exposure_time = self.parameters
```
Comment (vilim) [inline sashimi/camera.py:155]: this does not make sense: you should pass numbers/actual values to the setters, not a whole paramters object

**Review by vilim — CHANGES_REQUESTED** [review]
I guess the mock_camera requires binning still (to test everything), and perhaps it would be good to generate only one picture at the beginning (it can be modulated every frame by multiplying it), binning can be done with scikit-image `block_reduce`, and then many things can be tested without having a hamamatsu camra (also the frame time dispatching is not exactly right, I suggest just sleeping a very small amount and checking time differences from last frame dispatched by using time_ns)

**Comment by vilim** [comment]
I will merge this only once the Hamamatsu camera is in, it doesn't make sense to have it on the master branch otherwise

**Comment by diegoasua** [comment]
Even though the wrapper is now looking better I am truly unhappy. It is ugly. At least now it looks legible, more than before. The functions heredated from the previous potpurri are long and weird and the initialization is hideous.

File: sashimi/hardware/cameras/__init__.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    @property
+    def sensor_resolution(self):
+        return self._sensor_resolution
+
+    @property
+    def frame_shape(self):
+        return tuple([None, None])
+
+
+class AbstractCameraConfigurator:
```
Comment (vilim) [inline sashimi/hardware/cameras/__init__.py:97]: This context manager is not being used

**Review by vilim — CHANGES_REQUESTED** [review]
Looks good! I think the ugliness of the Hamamatsu wrapper is mostly unavoidable. Check now if you can get a GUI running with the MockCamera, and then I try to merge it on the scope when I get the chance (I can also do it remotely as long as the camera is on)

**Comment by diegoasua** [comment]
Tested in my laptop with `--scopeless` mode everything seems fine

File: sashimi/camera.py [hunk 13]
```diff
@@ -124,158 +128,42 @@ def run(self):
...
-                ):
-                    break
-            except Empty:
-                pass
+    def update_parameters(self):
+        self.parameters = self.new_parameters
+        for attribute in ["exposure_time", "binning", "roi", "trigger_mode"]:
+            setattr(self.camera, attribute, getattr(self.parameters, attribute))
+        # the following is to match internals of the camera as some cameras are restricted (e.g. only multiples of 4)
+        self.parameters.frame_shape = self.camera.frame_shape
```
Comment (vilim) [inline sashimi/camera.py:161]: I think this is also the case for exposure. But these parameters are (right now) not sent to the main process, right?

File: sashimi/utilities.py [hunk 14]
```diff
@@ -33,3 +35,36 @@ def get_last_parameters(parameter_queue, timeout=0.0001):
...
+    Buffer for large data arrays based on numpy and using ctypes for speedy copy of data
+    """
+
+    def __init__(self, size=None, *args, **kwargs):
+        """
+        Create a data object of the appropriate size.
+        """
+        super().__init__(**kwargs)
+        self.np_array = np.ascontiguousarray(
+            np.empty(size // 2)
```
Comment (vilim) [inline sashimi/utilities.py:52]: why did you remove the dtype here? I think it's necessary, this initializes float64 arrays whereas cameras use uint16

File: sashimi/utilities.py [hunk 15]
```diff
@@ -33,3 +35,36 @@ def get_last_parameters(parameter_queue, timeout=0.0001):
         except Empty:
             break
     return params
+
+
+# TODO: Consider this for ScopeCuisine
+class SpeedyArrayBuffer:
```
Comment (vilim) [inline sashimi/utilities.py:41]: I would call it PointerArray or CPointerArray

File: sashimi/utilities.py [hunk 16]
```diff
@@ -33,3 +35,36 @@ def get_last_parameters(parameter_queue, timeout=0.0001):
...
+        super().__init__(**kwargs)
+        self.np_array = np.ascontiguousarray(
+            np.empty(size // 2)
+        )
+        self.size = size
+
+    def __getitem__(self, slice):
+        return self.np_array[slice]
+
+    def copy_data(self, address):
```
Comment (vilim) [inline sashimi/utilities.py:59]: I would rename it into `copy_into` (you are copying the data from the address to the array, not vice-versa, right?)

**Review by vilim — CHANGES_REQUESTED** [review]
Looking good! They are some minor changes, but I can fix them myself along with formatting on merge.

File: sashimi/camera.py (hunk 13 above)
Comment (diegoasua) [reply sashimi/camera.py:161]: In 'camera_gui' there is a function to show the user the current shape of the frames:

```python    
    def update_roi_info(self, width, height):
        self.lbl_roi.setText(
            "Current frame dimensions are:\nHeight: {}\nWidth: {}".format(
                int(height), int(width)
            )
        )
```

Now, this function is actually wrong because it is diplaying height and width which are some legacy parameters from previous versions. It should be changed so that it takes intead the `frame_shape` tuple and displays it. This can be done in this PR but also can be done in a further one dealing with #87 .

Comment (vilim) [reply sashimi/camera.py:161]: Ok!

**Comment by vilim** [comment]
I think the linter will not be happy until you remove the `from __ import *` thing

**Comment by vigji** [comment]
maybe with such a long importing the best practice would be to import the module as some 3 letters shortening

**Comment by diegoasua** [comment]
> maybe with such a long importing the best practice would be to import the module as some 3 letters shortening

For each imported class that may be a bit too much. Plus the naming is informative of what is happening. One can just collapse the import section

**Comment by vilim** [comment]
@diegoasua I wanted to merge this, but setting a ROI in with the mock camera crashes sashimi.

**Comment by diegoasua** [comment]
Let me check that now

**Comment by diegoasua** [comment]
I think I found the issue.

It is this line of [code](https://github.com/portugueslab/sashimi/blob/64bc0e0ae536004161cf375e02086d0445a0f176/sashimi/hardware/cameras/mock.py#L69) in `np.block_reduce()`. One can perfom the pool down by applying a function like `np.sum()` or `np.mean()`. Because we are working with signed unintegers and using here `np.sum()` it may commit overflow in some cases.

**Comment by vilim** [comment]
hmm yes, it doesn't always fail. But it definitely behaves strangely:

**Comment by vilim** [comment]
So I don't think the issue for block_reduce (you can always set the function to np.max or np.min to avoid overflow though)

**Comment by diegoasua** [comment]
Sorry for this issues that I should had spotted before committing and thanks @vilim for bringing them to my attention. This is resolved for our renewed frisky `MockCamera`. Also fixed other unrelated issues that found while searching for the problem. Can't say with 100% confidence this will work for Hamamatsu without further modification, as it is now -- (as I said when I opened this PR the chances of messing up the actual camera were high). Specifically it is not clear to me what would happen if the sCMOS receives unexpected shapes, e.g. a roi bigger partially overlapping with the image. For the `MockCamera` this was solved by [pre-processing with a simple cropper](https://github.com/portugueslab/sashimi/blob/26dc1d8c5cd15b28ec282183ebab3f9362883c60/sashimi/hardware/cameras/mock.py#L56).

One thing I want to bring attention to, is the fact that one can blow up the whole software if the ROI shape gets moved by the corner handles and unfortunately ends up upside down or laterally inverted, raising `OverflowError: cannot convert float infinity to integer` fatal error due to the roi having negative size(s). To correct this one should either hard-code a fix to revert/fix dimensions of ROI or simpley we should improve the UX with the ROI selection (which includes other aspects appart from this one)

**Comment by vilim** [comment]
The solution to all these questions is simple: you preprocess the coordinates you get from the GUI before setting the ROI, regardless of camera (this should happen in the GUI code, not camera code):
1) sorting the coordinates
2) taking the intersection between the ROI coordinates and the displayed image coordinates 
(for one of them it's x_min_new = max(x_min, 0)), x_max_new = min(im_w, x_max))

**Comment by diegoasua** [comment]
There must be another PR dealing with ROI selection. It is almost harrowing, distressful...

**Comment by vilim** [comment]
I almost got it to work with the real camera, however there is some weird pause on changing any camera settings, overfilling the buffer. Also, the ROI interface is worse now than before: a horrible yellow square is visible at start, and you actually can't see the whole image until the ROI is set. Making the square yellow also does not help with finding the almost-invisble handles at the edges and just obscures what you see. The ROI interface should instead have 3 separate texts to reflect 3 states: 
1) full frame is set: select ROI, 
2) ROI square is displayed: set ROI, cancel <- there are 2 buttons here and only in this state 
3) an ROI is set: reset full frame

**Comment by diegoasua** [comment]
What I wonder is whether we should address the ROI problems here or in a separate PR

**Comment by vilim** [comment]
Since you already started making changes to the ROI workflow here, you have to either revert them or bring them to a more usable state.
The 3-state thing should be easy to implement, fixing the box visibility will require monkey-patching around [here](https://github.com/napari/napari/blob/bb5befa36e3f2e894fa42af48465e448f9c1dcda/napari/layers/shapes/shapes.py#L884), but this can indeed be a separate PR.
Once the ROI is fixed, I will figure out what the issue is with the camera parameter settings or get Luigi to do it, it's not possible without the camera attached.

**Comment by diegoasua** [comment]
To be able to change face color we should override [this in napari](https://github.com/napari/napari/blob/bb5befa36e3f2e894fa42af48465e448f9c1dcda/napari/utils/colormaps/_tests/colors_data.py#L116) because null color is declared as invalid, and it actually warns you and resets the color if you try to set `face_color` to null:

>UserWarning: The provided face_color parameter contained illegal values, resetting all face_color values to white.

Ideas on how to patch this welcome. Other than that I believe overall the ROI selection experience has improved significatively

**Comment by vilim** [comment]
I still did not manage to find what is wrong. Somehow setting the camera settings (everything that should call set_property_value) has no effect, even though no errors are raised or anything. There is something missing then in the property setting/acquistion start/other part of the cycle, but I couldn't figure out yet what it is.

**Comment by vilim** [comment]
A small point on the ROI: I think you can tell Napari to make it selected when it appears, so you can immediately see the handles.

**Comment by diegoasua** [comment]
I have no idea either what can be causing this. As I have no access to the camera I can't debug it either... Unless someone hooks me up through a desk sharing app, if that is possible, and leaves the camera on. For what I can see in the code everything seems fine. The fact that the `MockCamera` works fine when changing settings points to some failure directly in the wrapper... Where? No idea. The `set_property_value()` function has not changed a lot from what it was, other than f-string refactoring and replacing prints for warnings.

File: sashimi/processes/camera.py [hunk 17]
```diff
@@ -167,15 +165,21 @@ def camera_loop(self):
                     self.image_queue.put(frame)
                     self.was_waiting = is_waiting
                     self.update_framerate()
-            try:
-                new_parameters = self.parameter_queue.get(timeout=0.001)
+
+            # Empty parameters queue and set new parameters with the most recent value
```
Comment (diegoasua) [inline sashimi/processes/camera.py:163]: This is a great idea for increased efficiency! Well done

**Comment by diegoasua** [comment]
Turning this into a proper PR for review

**Comment by diegoasua** [comment]
The user guide editor does not work well. I have tried to fix it but I was unable, will appreciate some feedback on that. Basically its a double column window with "raw"markdown and "rendered" markdown respectively. The raw one is not really displaying raw, so when converted to rendered it can fail a bit. So the question, I guess, is how to show unrendered markdown (e.g. ##Title instead of __Title__) using `markdown` python in-built library or an external library.

**Comment by diegoasua** [comment]
Also: Is Travis not running?

**Comment by vigji** [comment]
> The user guide editor does not work well. I have tried to fix it but I was unable, will appreciate some feedback on that. Basically its a double column window with "raw"markdown and "rendered" markdown respectively. The raw one is not really displaying raw, so when converted to rendered it can fail a bit. So the question, I guess, is how to show unrendered markdown (e.g. ##Title instead of Title) using markdown or an external library.

I have not looked into that, but having an in-software editor markdown sounds like so off-topic that I would not have it in sashimi at all!

**Comment by vigji** [comment]
In general, I think this is ready for @vilim to have a look. I merged master here already, and it seems to be running (wit the same amount of bugs that master currently has).
A few general considerations:
 - keeping binning & poi size together might be tricky, that's the reason for some choices (eg connecting camera params changes to a method of the GUI and not of the State);
 - debugging this it seems that until we try a different camera, not sure of how much we can call our abstract class really abstract. Actually, I did the debugging working mostly on the side of the Ham Interface, and not that much on the abstract class, apart from moving a method there. Let me know if you have other suggestions;
- now the config file is used to set default binning and exposure.
- the camera is always stopped and started around the parameters setting.

The update of the view itself seems to run very smoothly

**Comment by diegoasua** [comment]
We were previously observing delays of ~80 ms for experiment trigger. I wonder whether that had something to do with the way the parameters were set (and so why we needed the 3GB queue). Is the delay still 80 ms now?

File: sashimi/config.py [hunk 18]
```diff
@@ -20,6 +20,11 @@
     "scanning": "mock",
     "scopeless": False,
     "sample_rate": 40000,
+    "voxel_size" : {
+        "x": 0.3,
+        "y": 0.3,
+        "z": 6.5
```
Comment (vilim) [inline sashimi/config.py:26]: hmm, strange to have z voxel size here, it is determined by spacing between planes

File: requirements.txt [hunk 19]
```diff
@@ -12,4 +12,6 @@ toml
 scopecuisine>=0.1.2
 pyqtgraph
 napari
-numba
\ No newline at end of file
+numba
+numba
```
Comment (vilim) [inline requirements.txt:16]: double numba?

File: sashimi/foo.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
+from sashimi.hardware.cameras.hamamatsu.interface import HamamatsuCamera
```
Comment (vilim) [inline sashimi/foo.py:1]: what's this file?

File: sashimi/gui/camera_gui.py [hunk 21]
```diff
@@ -47,237 +61,260 @@ class ViewingWidget(QWidget):
...
         self.bottom_layout = QHBoxLayout()
-        self.bottom_layout.addWidget(self.wid_display_settings)
-        self.bottom_layout.addWidget(self.btn_display_roi)
-        self.bottom_layout.addWidget(self.btn_reset_contrast)
+
+        self.viewer.window.qt_viewer.viewerButtons.consoleButton.hide()
+        self.viewer.window.qt_viewer.viewerButtons.rollDimsButton.hide()
+        self.viewer.window.qt_viewer.viewerButtons.transposeDimsButton.setText("Rotate view")
+        self.viewer.window.qt_viewer.viewerButtons.resetViewButton.setText("Reset view")
+        self.viewer.window.qt_viewer.viewerButtons.gridViewButton.setText("Grid mode")
```
Comment (vilim) [inline sashimi/gui/camera_gui.py:102]: The grid mode dosn't do anything useful for the ligthsheet

File: sashimi/gui/camera_gui.py [hunk 22]
```diff
@@ -47,237 +61,260 @@ class ViewingWidget(QWidget):
...
+    def auto_contrast(self):
+        if self.auto_contrast_chk.isChecked():
+            self.wid_contrast_range.hide()
+            self.auto_contrast = True
+        else:
+            self.wid_contrast_range.show()
+            self.auto_contrast = False
+
+    def set_manual_contrast(self):
+        self.frame_layer.contrast_limits = [i for i in self.contrast_range.contrast_range]
```
Comment (vilim) [inline sashimi/gui/camera_gui.py:188]: why the list comprehension? what is .contrast_range so that it has to be converted in this involved way?

File: sashimi/gui/camera_gui.py [hunk 23]
```diff
@@ -47,237 +61,260 @@ class ViewingWidget(QWidget):
...
-
-        self.set_roi_button.clicked.connect(self.set_roi)
-        self.set_full_size_frame_button.clicked.connect(self.set_full_size_frame)
+        self.layout().addWidget(self.btn_roi)
+        self.layout().addWidget(self.btn_cancel_roi)
+
+        self.btn_roi.clicked.connect(self.roi_action)
+        self.btn_cancel_roi.clicked.connect(self.cancel_roi_selection)
+
+    def roi_action(self):
```
Comment (vilim) [inline sashimi/gui/camera_gui.py:233]: hmm this cycles through ROI states, maybe could be named better?

Comment (vilim) [reply sashimi/gui/camera_gui.py:233]: and separated into two (iterate state and apply state change)

File: sashimi/gui/camera_gui.py [hunk 24]
```diff
@@ -47,237 +61,260 @@ class ViewingWidget(QWidget):
...
+
+    def _hide_roi(self):
+        self.wid_display.roi.visible = False
+
+    def _show_roi(self):
+        self.wid_display.roi.visible = True
+
+    def cancel_roi_selection(self):
+        self.roi_state = RoiState(3)
+        self.roi_action()
```
Comment (vilim) [inline sashimi/gui/camera_gui.py:270]: see comment about the function above

File: sashimi/gui/status_bar.py [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+
+            # Find the expected framerate depending on the global state
+            if self.state.global_state == GlobalState.PREVIEW:
+                expected_frame_rate = None
+            if self.state.global_state == GlobalState.VOLUME_PREVIEW:
+                vol_sets = self.state.volume_setting
+                n_planes = (vol_sets.n_planes - vol_sets.n_skip_start - vol_sets.n_skip_end)
+                expected_frame_rate = vol_sets.frequency * n_planes
+            if self.state.global_state == GlobalState.PLANAR_PREVIEW:
+                expected_frame_rate = self.state.single_plane_settings.frequency
```
Comment (vilim) [inline sashimi/gui/status_bar.py:69]: this calculation doesn't seem like something that should be in the GUI code (rather state)

File: sashimi/hardware/scanning/scanloops.py [hunk 26]
```diff
@@ -85,6 +85,12 @@ class ScanParameters:
 
 
 class ScanLoop:
+    """General class for the control of the event loop of the scanning, taking
+    care of the synchronization between the galvo and piezo scanning and the camera triggering.
+    It has a loop method which is defined only here and not overwritten in sublasses, which controls
+    the main order of events. In this class we control only the lateral scanning, which is common to both
+    planar and volumetric acquisitions.
+    """
```
Comment (vilim) [inline sashimi/hardware/scanning/scanloops.py:93]: nice explanation

File: sashimi/processes/dispatcher.py [hunk 27]
```diff
@@ -31,37 +34,38 @@ def __init__(
...
         self.logger.log_message("started")
         while not self.stop_event.is_set():
             self.send_receive()
             self.get_frame()
         self.close_log()
 
-    def process_frame(self):
-        if self.current_frame is not None:
+    def process_frame(self, current_frame=None):
+        if current_frame is not None:
```
Comment (vilim) [inline sashimi/processes/dispatcher.py:53]: "flat is better than nested": invert the condition and just return immediately if it is None

**Review by vilim — CHANGES_REQUESTED** [review]
Some code comments, will do a UX review at some point later. Happy to see this progress :)

File: sashimi/gui/status_bar.py (hunk 25 above)
Comment (vigji) [reply sashimi/gui/status_bar.py:69]: but the frame rate is calculated for displaying purposes only, that is the reason this was here I guess. You still stand behind the point?

File: sashimi/gui/camera_gui.py (hunk 23 above)
Comment (vigji) [reply sashimi/gui/camera_gui.py]: not sure what the advantage would be in this case

**Comment by vilim** [comment]
I tried to trial the current version, but I get this error
```Traceback (most recent call last):
  File "/home/vilimstih/miniconda3/envs/sashimi/bin/sashimi", line 33, in <module>
    sys.exit(load_entry_point('sashimi', 'console_scripts', 'sashimi')())
  File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/vilimstih/phd/sashimi/sashimi/main.py", line 24, in main
    main_window = MainWindow(st)
  File "/home/vilimstih/phd/sashimi/sashimi/gui/main_gui.py", line 41, in __init__
    self.wid_display = ViewingWidget(st, self.timer)
  File "/home/vilimstih/phd/sashimi/sashimi/gui/camera_gui.py", line 88, in __init__
    self.viewer.theme = "dark_sashimi"
  File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/napari/components/viewer_model.py", line 144, in theme
    raise ValueError(
ValueError: Theme 'dark_sashimi' not found; options are ['dark', 'light'].
```

**Comment by diegoasua** [comment]
> I tried to trial the current version, but I get this error
> 
> ```
>   File "/home/vilimstih/miniconda3/envs/sashimi/bin/sashimi", line 33, in <module>
>     sys.exit(load_entry_point('sashimi', 'console_scripts', 'sashimi')())
>   File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/home/vilimstih/phd/sashimi/sashimi/main.py", line 24, in main
>     main_window = MainWindow(st)
>   File "/home/vilimstih/phd/sashimi/sashimi/gui/main_gui.py", line 41, in __init__
>     self.wid_display = ViewingWidget(st, self.timer)
>   File "/home/vilimstih/phd/sashimi/sashimi/gui/camera_gui.py", line 88, in __init__
>     self.viewer.theme = "dark_sashimi"
>   File "/home/vilimstih/miniconda3/envs/sashimi/lib/python3.8/site-packages/napari/components/viewer_model.py", line 144, in theme
>     raise ValueError(
> ValueError: Theme 'dark_sashimi' not found; options are ['dark', 'light'].
> ```

That's because we have a forked version of napari. Uninstall your current napari and run:
```
pip install git+https://github.com/portugueslab/napari.git#egg=napari
```

**Comment by vigji** [comment]
yeah that should at lest be made optional

**Comment by diegoasua** [comment]
Remember now to uninstall the forked napari from your local environment and install the latest version of it (currently 0.4.0). Also a side note for the future: If you pull a new version you might have to delete `user\.sashimi` as the template config file might have changed - pointing this as it has also happened to myself

**Comment by diegoasua** [comment]
@vigji it seems that the ROI is flipped? i.e. if one sets an horizontal rectangle in the mock I get it flipped. Does this happen both in `Mock` and actual camera or only in `Mock`?

**Comment by vilim** [comment]
The windows opens! And looks much better and more intuitive than before :) Can someone of you check though that `sashimi --scopeless` works with deleted hardware_config.toml ? Or provide a version of it for testing?

**Comment by vigji** [comment]
> @vigji it seems that the ROI is flipped? i.e. if one sets an horizontal rectangle in the mock I get it flipped. Does this happen both in `Mock` and actual camera or only in `Mock`?

Had to change the orientation to follow convention for images (first dimension=height, second dimension = width); there was a double inversion in passing the ROI to the camera that was making it working, so my guess is that mock camera should be fixed!

**Comment by diegoasua** [comment]
> The windows opens! And looks much better and more intuitive than before :) Can someone of you check though that `sashimi --scopeless` works with deleted hardware_config.toml ? Or provide a version of it for testing?

If you run it in a machine that does not have the config file it automatically creates a directory and a template file on first run

**Comment by diegoasua** [comment]
`MockCamera` is not pausing when entering into `CameraMode.PAUSED` mode. Does this only happen with the mock or also with the actual camera?

**Comment by vigji** [comment]
> After moving to calibration mode, MockCamera is not pausing when (supposedly) entering into CameraMode.PAUSED mode. Does this only happen with the mock or also with the actual camera?

I can't tell you for sure that the camera stops (at least in the view) when you are in the Paused mode. Is this what you mean?

**Comment by diegoasua** [comment]
> > After moving to calibration mode, MockCamera is not pausing when (supposedly) entering into CameraMode.PAUSED mode. Does this only happen with the mock or also with the actual camera?
> 
> I can't tell you for sure that the camera stops (at least in the view) when you are in the Paused mode. Is this what you mean?

Yes that is what I mean. The Mock Camera keeps running when you move to PAUSED and I just wanted to know whether it was something due to `mock.py` or a more general problem in `processes\camera.py` or `state.py`

**Comment by vilim** [comment]
Coming along very nicely! Some notes: camera exposure of mockcamera can go to 0, crashing the program. File -> save as something does not do anything and freezes the program. The config editor would work better as a two-column layout, with a list of settings on the left and editing box on the right (in the future a multi-column layout for hieararchical editor). Markdown editor for instructions is definitely out of scope. The roi when it appears is not selected and it's quite unintive to get the handles -> can it be made selected programatically? The mock picture of the camera could be something nicer and more slowly changing, this is a bit epileptic, also it's unclear what happens when you select a ROI. Maybe Luigi can provide a small blinking loop of a part of zebrafish brain (can be e.g. 256x256x10). Starting the experiment is a bit confusingly a button in the scanning settings, very far from the saving settins: I suggest it to be a prominent button, disabled unless you are in one of the two valid experiment states (Planar or Volume)

**Comment by diegoasua** [comment]
I am trying to reproduce what happened with file save in.. (I understand you mean save in... not save as...). For me it works nicely and does not produce any errors. Can you give more details about that?

**Comment by vilim** [comment]
What is "Save in..." even supposed to do? Determine the folder? Maybe it's a weird Linux bug

**Comment by diegoasua** [comment]
It is meant to set the folder where to save the split dataset.

**Comment by vilim** [comment]
I would call it "choose data destination folder" because you have the settings just before, so it's slightly confusing

**Comment by vigji** [comment]
@diegoasua if you agree I would add last remarks from @vilim as issues and merge this to master, as it is anyway what is in use at the lightsheet at the moment.

**Comment by diegoasua** [comment]
Good to see this being merged, sashimi is starting to take off! Good work both
